### ISSUE TITLE
Rethink apsim version numbers

### ISSUE THREAD

**Issue by hol430** [APSIMInitiative/ApsimX#6885]
We need to have a think about how/when we generate releases, and how we construct version numbers. There are multiple problems here and everyone seems to have different expectations.

## Background

The apsim version numbers are currently generated when a pull request is tested. The version number has the following format:

```
yyyy.MM.dd.$issue
```

where the date (`yyyy.MM.dd`) is the current date, and `$issue` is the issue number referenced by the pull request which is being tested. The version number is used in multiple places:

- Upgrades area of the GUI, and the downloads webpage
- Displayed in the files generated by autodocs
- Branded on the assemblies (ApsimNG and Models), which causes it to show up in the GUI (title bar) and when executing `Models.exe --version`

## Problems

Previously, the installers and autodocs were generated after pull requests were merged, so the date component of the version number came from the date on which the pull request was merged. This was a problem because changes introduced by a pull request could cause the autodocs to break, which we wouldn't realise until the autodocs were generated (after the pull request had already been merged). To solve this problem, we have now started generating the autodocs (and installers) when the pull request is tested (usually immediately after the pull request is created), which presents its own set of problems.

The date component of the version number is now the date on which the pull request is tested, which is often not the same as the date on which the pull request is merged. This means that if pull requests are not merged in the same order in which they were created, the upgrades' version numbers will appear to be out-of-order (in terms of the date component).

Another problem with the version numbers is the use of an issue number:

- These are often out-of-order relative to the release order, which leads to a lot of confusion
- Theoretically, an issue can be "fixed" multiple times if it's re-opened, allowing multiple releases to share an issue number in their version component

Another problem is that when multiple pull requests are merged in quick succession, the installers for these releases will not contain each others' changes. This isn't related to the version number, but is a (perhaps minor) problem with our current continuous release cycle.

There may be other problems which I've forgotten. Would be interested to hear others' thoughts here.

**Comment by zur003** [comment]
Would it be helpful to use the pull-request number rather than the issue number? This would generally be closer to the release order.

**Comment by jbrider** [comment]
Can the part of the autodocs that depend on the version be inserted once the release is made?

Can the release process be put into a 'holding pattern' for 'x' hours to allow for multiple pull requests in a day to be merged into a single release?  A good default would be to do it once a night?

I'm aware that there is a requirement to force a release in some cases - but that could just set the holding time to 0.

**Comment by hol353** [comment]
There are 2 main problems:

1. Using the issue number in the version string is confusing.
2. Some users think there are too many releases.

A solution:

1. Replace the issue number in the version string with a build number that simply increments. There should already be a number we can use from the builds database.
2. Add a filter selection to the upgrades view in the GUI and the downloads web page that lets the user choose 'weekly', 'monthly', 'yearly', 'all releases' (default: monthly?). The users choice in the GUI should be remembered. The tricky part of this would be to show a complete list of changes for a version when the user chooses something other than 'all releases'.

This lets us keep the build system as it is (which works well for the development team). It lets the user choose the level of detail they are comfortable with. It doesn't change our git branching model which I know our less experienced git users would have a problem with.

This doesn't solve all the issues outlined above. Thoughts?

**Comment by jbrider** [comment]
These both sound good.

**Comment by hol430** [comment]
Incrementing version number sounds good and would solve the confusion around the version numbers which currently don't always go up.

> Add a filter selection to the upgrades view in the GUI and the downloads web page that lets the user choose 'weekly', 'monthly', 'yearly', 'all releases' (default: monthly?)

I'm not violently opposed to this, but I also don't really see the point - I think it would be less confusing and simpler to just move to a less regular (more irregular) release cycle (ie weekly/monthly/centennial...). The branching model mentioned by Rogerio is a good way to do this, but if you think it would be confusing, we could just make irregular stable/official releases and tell people to build from source if they want the absolute latest changes.

On the other hand, if people want less upgrades, they could just upgrade less regularly. If we want to make it easier for people to upgrade to a specific version, we could add a search box to the upgrades area of the GUI and the website. I do think there's something to be said though for having less regular but more stable releases, in which we have some confidence that things are actually working.

**Comment by lie112** [comment]
As ASPIM Next Gen is a complete rebuild of Classic and does not have years of bug fixes behind it, the user will benefit from many bug fixes and important improvements especially during active development of some models. This has been a very powerful benefit of the current continuous release approach. CLEM users reporting a bug preventing their work can have this fixed and a new version available usually within 24 hours (given fast pull approval) without needing to be using a development environment, which 99.9% would not cope with. Some of my users are already developing "upgrade anxiety" as they have the belief that an upgrade will lead to more problems or require them to make changes to their simulation... This is not entirely true (but has happened with large refactoring to give them so much more functionality), and they really need to improve their skills and not be afraid of following the instructions in a warning or error to get things working.

What about less regular upgrades reported in the upgrade list, but the ability for users to enter a specific commit number to upgrade to that version? I usually report the commit upgrade number in my [help system](https://www.apsim.info/clem/Content/Details/WhatsNew.htm). This would allow them to upgrade to the latest CLEM changes and not need any other upgrades. A full list of the upgrades on a website or the interface would be useful, maybe with some tags for filtering (see below). I DO NOT support providing less upgrades and moving to dev environment for users that need to be kept up with the most recent release.

A nested tree of year, month, all releases in the month would be useful with the year and month nodes linking to the latest build of that time period.

The problem is we have too much information and choice for the user. Ideally, our system would show upgrades based on each model, interface, core of APISM, and other tools. Then users can see what aspects are being worked on. It would be nice to flag upgrades. Critical upgrades would show where it is really worth upgrading for major fixes. Model based upgrades would show where the model they are interested in has been improved (the model should be associated with the pull request). But this all requires upgrade functionality code changes. Other categories could be UserInterface.

Maybe we need an information tab or block of text explaining the advantages of the current upgrade process in the upgrade window and how we will let them grab the latest weekly, monthly or annual by selecting that entry in the tree. 

1. A monthly (or continuously) incrementing version number would avoid confusion of the 4 digit number that isn't always in ascending order. 

I had always assumed that each build would contain all previous commits and so was the latest version as mentioned in @hol430 edit:

**Comment by hol353** [comment]
>  think it would be less confusing and simpler to just move to a less regular (more irregular) release cycle (ie weekly/monthly/centennial...)

Agreed but we sometimes provide bug fixes for non technical users that 'need it now'. We don't want them to be using git and visual studio - too hard. Less frequent releases makes quick turn arounds for these people impossible.

> On the other hand, if people want less upgrades, they could just upgrade less regularly. 

Yep agree totally!

> I do think there's something to be said though for having less regular but more stable releases

Yes, but how do we determine which releases are more 'stable'? In APSIM Classic we would test a build for releaseability. It took ages and we don't have the resources to do that.

@lie112 suggests some interesting ideas about a hierarchical tree of upgrades with tags, filtering, searching. This would require quite a lot of time to implement and ironically give the user more information and choice on how to find a particular release.

I'm trying to find the simplest possible way forward.

**Comment by hol430** [comment]
> As ASPIM Next Gen is a complete rebuild of Classic and does not have years of bug fixes behind it

Old apsim also had years of bugs behind it :)

If we're not going to guarantee stability of irregular releases (e.g. via manual testing) then you're right - there's probably not really any advantage in reducing the number of releases, or changing to the branching model mentioned by Rogerio. Of course, that still leaves us with the problem of releases not always containing changes from previous releases. What other problems are we trying to resolve here? E.g. you both mentioned filtering/metadata for the releases in the gui - what problem are we trying to solve here?

**Comment by hol353** [comment]
I only mentioned filtering as a way of reducing the number of releases the user sees. Personally, I don't want filtering.

> Of course, that still leaves us with the problem of releases not always containing changes from previous releases

The only way to fix this problem is to enforce all PRs to be at the tip revision. I think GitHub can enforce this. Once a PR is merged then any other open PRs will be automatically invalidated and their owner will need to do a pull and then a push to get them up to date. This would be a pain at busy times like a sprint.

**Comment by jbrider** [comment]
When a pull request is merged, I think it might be possible to automate the update of other pull requests to the tip version. I'm not sure if our build system would allow this though.  

As mentioned above, I don't think the number of updates is the real issue but the frequency that the user is made aware of them?  At the client end, we could enable a weekly/monthly check so it doesn't automatically check every time they run it.  Manually calling check for updates would still look for the most recent.  

You could also add a tag to the updates that indicates importance ie: (Critical, Important, Regular, Please don't bother me) which the user could select what level they would be notified at. I would think the reality of this would be that most would be regular with the very occasional Critical that fixes a breaking bug.

**Comment by hol430** [comment]
Current plan:

- Incrementing version number (yyyy.MM.X.0)
- Version number starts out at same as current version (+1)
- Remove version # from autodocs, but include it in autodocs file names
- Build installers + docker images after PR is merged

**Comment by hol430** [comment]
The above have all been implemented now, so I'm going to close this as fixed. I'll open new issues for any new bugs which pop up as a result of my "improvements".
